drivers/adc/ad7616/iio_ad7616.c

```diff
@@ -0,0 +1,310 @@
+ */
+static int32_t iio_ad7616_debug_reg_write(void *device, uint32_t reg,
+		uint32_t val)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	return ad7616_write(dev, (uint8_t)reg, (uint16_t)val);
+}
+
+static int get_raw(void *device,
+		   char *buf,
+		   uint32_t len,
+		   const struct iio_ch_info *channel,
+		   intptr_t id)
+{
+	int ret;
+	uint8_t index = channel->ch_num < 8 ? channel->ch_num : channel->ch_num - 8;
+	struct ad7616_conversion_result res[8];
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	// TODO: we call this on every call, fetching all 16 inputs. Doesn't look smart
+	ret = ad7616_read_data_serial(dev, res, 1);
+	if (ret)
+		return ret;
+
+	if (channel->ch_num < 8)
+		return sprintf(buf, "%i", res[index].channel_a);
+
+	return sprintf(buf, "%i", res[index].channel_b);
+}
+
+static int get_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	enum ad7616_range range = (ch < AD7616_VB0) ? dev->va[ch] : dev->vb[ch -
+				  AD7616_VB0];
+
+	switch (range) {
+	case AD7616_10V:
+		return sprintf(buf, "10");
+	case AD7616_5V:
+		return sprintf(buf, "5");
+	case AD7616_2V5:
+		return sprintf(buf, "2.5");
+	default:
+		return -EINVAL;
+	}
+}
+
+static int set_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	int ret;
+	int val;
+	enum ad7616_range range;
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	ret = sscanf(buf, "%u", &val);
+	if (ret != 1)
+		return -EINVAL;
+
+	switch (val) {
+	case 10:
+		range = AD7616_10V;
+		break;
+	case 5:
+		range = AD7616_5V;
+		break;
+	case 2:
+		range = AD7616_2V5;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ad7616_set_range(dev, ch, range);
+}
+
+/**
+ * @brief	Read buffer data corresponding to AD7616 IIO device
+ * @param	iio_dev_data[in] - Pointer to IIO device data structure
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int32_t iio_ad7616_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+	struct ad7616_conversion_result results[8];
+	struct ad7616_iio_dev *iio_dev = iio_dev_data->dev;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+	uint16_t data[iio_dev_data->buffer->bytes_per_scan / 2];
+	int32_t ret;
+	uint32_t i;
+	uint32_t j;
+	uint32_t k;
+
+	for (i = 0; i < iio_dev_data->buffer->samples; i++) {
+		ret = ad7616_read_data_serial(dev, results, 1);
+		if (ret)
+			return ret;
+
+		for (j = 0, k = 0; j < 16; j++) {
```

Comment:
I can propose this:
![image](https://github.com/analogdevicesinc/no-OS/assets/15075443/9d9d29ee-9b35-43e7-b892-dece4266820a)

The built-in is supported by both gcc and clang (checked with godbolt) and we are iterating exactly on the number of requested channel

Reply:
please see my remark in the comment below... and please be careful when using ad7616_read_data_serial(dev, results, 1); as shown in the implementation if the number of samples is equal to 1 only results[0] will be written. results[index] with index anything higher than 0 will not contain data read from the device.

Reply:
wouldn't no_os_find_first_set_bit do the job instead of ffs?
the code is a little bit hard to follow.

Reply:
I would argue that the `__builtin` solution [looks](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYMQAZg2kHAGTwGTAA5dwAjTGIQACYuUgAHVAVCWwZnNw9vXySUmwFA4LCWSOi4i0wrfIYhAiZiAgz3Tx8KqrTa%2BoJC0Iio2PiFOoamrNahrp7i0oGASgtUV2Jkdg5XBhTgYPQAaiCCHZpiIeMlAmNwwgh1ze29wR2AdxJ0WZ2AUgB2ACF3jQBBHZAnY3PBbTC7faHPDHc5nC6ED5eAAiOw07y8vwBf0BwMeCDomB2EGexFeHx%2BOOB1L2VGJpN27xiADY0aouLMqTTucRMAQlgxobDTnyEQQMVjcdyudyGUjsBiFSidlwJTKaUcTvDLuKYr89WqAdSvsicerefziIKvDFDf8TWaAaDwZCHpq4aKdTFrhswXcoQy3l9JdTnf63TCtZ7ERjUaZwq46DYGMYqFQFCSXpzMY6pbTiWAwO6RecdZyjdydhaBTsbXbzXya8XtYiALQq%2BufZEcea0TgAVl4ng4WlIqE4jh2CkWyyJTK8PFIBE0PfmAGsQP3fH2OJIhyux5xeAoQL5lyOe6Q4LAYIgUKgWAlCWQKBA0I/nyh6ixmZJgFwbV8GhaAIKITwgcID0uZhiAAT04RdoPqWCAHlwm0TBrAQ3h3zYQQUIYWh4IvUgsATYBHDEWgT24XgsBYQxgHEEj8F5aw8AAN0wGjR0wVRMNcUCD32SoD1oPBwmIZDnCwA8CGIPAWGw%2BYqAMYAFAANTwTBHhQhJGGwmRBBEMR2CkIz5CUNQD10eIDCMEBTGMcxxPCE9IHmVAEmqGjWxQmIdlbAB1KjAqCvj5KYXhUC44gFKwdyIHmSxMOqewGCcFxmgkeIAiCXoSn6GJ%2B0SZJUgEUZPC4eJcnKhgpj6MoSpS9iBE6EYsqyaq2lSjphm6fLpiK5r%2BsqnKLH6hrCqa5KZxWCRewHfcSPHDgdn%2BAAlABZX8dgAcUcScAIAOhiY6NGJXBCBID4YgXWZeHPLRZnXTdt04PdSCUrdSGHUdVuPU8lxXeZrzvRYCASQTyHsfAiHivR%2BGM0RxHMpHLJUdQSNs0hHikhJlP0JbfoPVaUMEqGDlQOlNp2yR9sOlUvFO87iWcD96GIW77sekH5gQTAmCwaIkqJ3deG%2B3w/uio8LCBp7V1IDcfp3Lxlv%2B2WFZesWYnVmWOF5i9tdilI7EkIA%3D) faster. It could be a good idea to rework `no_os_find_first_set_bit` in the future. Anyway I will use it as it's more inline with the other projects. Thanks for pointing this out!